File: bms/go/geo/geo_r3_PreciseVector_Vector/geo_r3_PreciseVector_Vector.go
package test

import (
	"math"
	"math/big"
)

func (v geo_r3_PreciseVector) Vector() geo_r3_Vector {
	// The accuracy flag is ignored on these conversions back to float64.
	x, _ := v.X.Float64()
	y, _ := v.Y.Float64()
	z, _ := v.Z.Float64()
	return geo_r3_Vector{x, y, z}.Normalize()
}

func (v geo_r3_Vector) Normalize() geo_r3_Vector {
	n2 := v.Norm2()
	if n2 == 0 {
		return geo_r3_Vector{0, 0, 0}
	}
	return v.Mul(1 / math.Sqrt(n2))
}

func (v geo_r3_Vector) Norm2() float64 { return v.Dot(v) }

func (v geo_r3_Vector) Dot(ov geo_r3_Vector) float64 {
	return float64(v.X*ov.X) + float64(v.Y*ov.Y) + float64(v.Z*ov.Z)
}

func (v geo_r3_Vector) Mul(m float64) geo_r3_Vector { return geo_r3_Vector{m * v.X, m * v.Y, m * v.Z} }

type geo_r3_PreciseVector struct {
	X, Y, Z *big.Float
}

type geo_r3_Vector struct {
	X, Y, Z float64
}

File: translations_only/success/go/geo/geo_r3_PreciseVector_Vector/claude2/Hinted/geo_r3_PreciseVector_Vector.rs

use std::f64;

struct PreciseVector {
    x: i64,
    y: i64,
    z: i64,
}

struct Vector {
    x: f64,
    y: f64,
    z: f64,  
}

fn vector(v: PreciseVector) -> Vector {
    let x = v.x as f64;
    let y = v.y as f64;
    let z = v.z as f64;
    normalize(Vector { x, y, z })  
}

fn normalize(mut v: Vector) -> Vector {
    let norm2 = norm2(&v);
    if norm2 == 0.0 {
        return Vector { x: 0.0, y: 0.0, z: 0.0 };
    } else {
        mul(&mut v, 1.0 / f64::sqrt(norm2));
    }
    v
}

fn norm2(v: &Vector) -> f64 {
    dot(v, v)
}

fn dot(v: &Vector, ov: &Vector) -> f64 {
    v.x * ov.x + v.y * ov.y + v.z * ov.z
}

fn mul(v: &mut Vector, m: f64) {
    v.x *= m;
    v.y *= m; 
    v.z *= m;  
}

